dvc/remote/ssh/__init__.py

```diff
@@ -306,6 +308,10 @@ def cache_exists(self, checksums, jobs=None, name=None):
         if not self.no_traverse:
             return list(set(checksums) & set(self.all()))
 
+        # possibly prompt for credentials before "Querying" progress output
+        path_info = self.checksum_to_path_info(checksums[0])
+        self.ensure_credentials(path_info.host, path_info.user, path_info.port)
```

Comment:
Hm, seems like we could simply do
```
with self.ssh(self.path_info):
    pass
```
And that way we won't need `ensure_credentials` at all.

Reply:
what about https://github.com/iterative/dvc/blob/94cfb840a25292cf72e558242cd294023a01f17c/dvc/remote/ssh/__init__.py#L138-L149

which is unneeded for just ensuring credentials exist? Is it not worth worrying about?

Reply:
@casperdcl Mm, not sure I follow. What is concerning you about that piece of code?

Reply:
Aah, got it. You mean that we don't have to establish the connection to check that we have the password, right?

Reply:
You are right, I've missread the way `def ssh` got split.

Reply:
right; though I missed `self.path_info`. I presume I should use that instead of the hacky `self.checksum_to_path_info(checksums[0])`? As the pre-existing comments say, it's unfortunate that per-path credentials aren't supported.

Reply:
Ok, then one more thing:
```
        path_info = self.checksum_to_path_info(checksums[0])
```
is not really needed. We can use self.path_info instead
```
self.ensure_credentials(self.path_info.host, self.path_info.user, self.path_info.port)
```
Though, I would probably change it to just accept path_info like so
```
self.ensure_credentials(self.path_info)
```

Reply:
:P I answered first